chapter-06/lambdas/create_link/src/main.rs

```diff
@@ -0,0 +1,197 @@
+    use shared::core::ShortUrl;
+    use shared::core::UrlInfo;
+    use shared::core::UrlRepository;
+    use shared::core::UrlShortener;
+    use shared::url_info::UrlDetails;
+
+    use crate::function_handler;
+
+    #[derive(Debug)]
+    struct TestUrlRepository {
+        return_error: bool,
+    }
+
+    #[async_trait]
+    impl UrlRepository for TestUrlRepository {
+        async fn get_url_from_short_link(
+            &self,
+            short_link: &str,
+        ) -> Result<Option<String>, String> {
+            Ok(Some("https://google.com".to_string()))
+        }
+
+        async fn store_short_url(
+            &self,
+            url_to_shorten: String,
+            short_url: String,
+            url_details: UrlDetails,
+        ) -> Result<ShortUrl, String> {
+            if (self.return_error) {
+                Err("Failure storing".to_string())
+            } else {
+                Ok(ShortUrl::new(
+                    short_url,
+                    url_to_shorten,
+                    0,
+                    url_details.title,
+                    url_details.description,
+                    url_details.content_type,
+                ))
+            }
+        }
+
+        async fn list_urls(
+            &self,
+            last_evaluated_id: Option<String>,
+        ) -> Result<(Vec<ShortUrl>, Option<String>), String> {
+            Ok((
+                vec![ShortUrl::new(
+                    "12345689".to_string(),
+                    "https://google.com".to_string(),
+                    0,
+                    None,
+                    None,
+                    None,
+                )],
+                None,
+            ))
+        }
+    }
+
+    #[derive(Debug)]
+    struct TestUrlInfo {}
+
+    #[async_trait]
+    impl UrlInfo for TestUrlInfo {
+        async fn fetch_details(&self, url: &str) -> Result<UrlDetails, String> {
+            Ok(UrlDetails {
+                content_type: Some("text/html".to_string()),
+                title: Some("Google".to_string()),
+                description: Some("Test description".to_string()),
+            })
+        }
+    }
+
+    #[tokio::test]
+    async fn when_valid_link_is_passed_should_store_and_return_details() {
+        let mockUrlRepo = TestUrlRepository {
+            return_error: false,
+        };
+        let testUrlInfo = TestUrlInfo {};
+
+        let url_shortener = UrlShortener::new(Box::new(mockUrlRepo), Box::new(testUrlInfo));
+
+        let request = Request::builder()
+            .header("Content-Type", "application/json")
+            .body(Body::from("{\"url_to_shorten\": \"https://google.com\"}"))
+            .unwrap();
+
+        let result = function_handler(&url_shortener, request).await;
+
+        assert_eq!(result.is_ok(), true);
+
+        let data = result.unwrap().into_response().await;
+
+        assert_eq!(data.status(), 200);
+
+        let response_struct: Value = serde_json::from_slice(data.body()).unwrap();
```

Comment:
honest question: why `from_slice` and not `from_str`?

Reply:
You can pass `data.body()` directly into the from_slice function. If you use from_str you get a mismatched type error :) You'd need to manually parse the `str` body out of the `http::response::Response` body